File: 2021/erlang/src/day02.erl
-module(day02).

-export([run/1]).

run(Input) ->
    io:format("day02 part 1 => ~p~n", [part1(Input)]),
    io:format("day02 part 2 => ~p~n", [part2(Input)]).

parse_op(Line) ->
    [Op, Val] = string:split(Line, " "),
    {list_to_atom(Op), list_to_integer(Val)}.

part1(Input) ->
    Ops = lists:map(fun parse_op/1, Input),
    Res = run_ops1(Ops, #{horizontal=>0, depth=>0}),
    #{depth:=D, horizontal:=H} = Res,
    D * H.

run_ops1([], Acc) -> Acc;
run_ops1([{forward, Val}|Rest], #{horizontal:=H} = Acc) ->
    run_ops1(Rest, Acc#{horizontal=>H + Val});
run_ops1([{up, Val}|Rest], #{depth:=D} = Acc) ->
    run_ops1(Rest, Acc#{depth=>D - Val});
run_ops1([{down, Val}|Rest], #{depth:=D} = Acc) ->
    run_ops1(Rest, Acc#{depth=>D + Val}).

part2(Input) ->
    Ops = lists:map(fun parse_op/1, Input),
    Res = run_ops2(Ops, {0, 0, 0}),
    {D, H, A} = Res,
    D * H.

run_ops2([], Acc) -> Acc;
run_ops2([{forward, Val}|Rest], {D, H, A} ) ->
    run_ops2(Rest, {D + (Val * A), H + Val, A});
run_ops2([{up, Val}|Rest], {D, H, A}) ->
    run_ops2(Rest, {D, H, A - Val});
run_ops2([{down, Val}|Rest], {D, H, A}) ->
    run_ops2(Rest, {D, H, A + Val}).

File: 2021/js/lib/day02.js
'use strict';

const run = (s) => {
    const input = s.split('\n');
    console.log(`day02 part 1 => ${part1(input)}`); // 1507611
    console.log(`day02 part 2 => ${part2(input)}`); // 1880593125
};

const parseOp = (line) => {
    const [cmd, val] = line.split(' ');
    return [cmd, parseInt(val)];
};

/**
 * @example part1([ 'forward 5', 'down 5', 'forward 8', 'up 3', 'down 8', 'forward 2' ])
 * //=> 150
 */
const part1 = (input) => {
    const ops = input.map(line => parseOp(line));
    const { h, d } = ops.reduce((acc, [cmd, val]) => {
        if (cmd === 'forward') return { ...acc, h: acc.h + val };
        if (cmd === 'down') return { ...acc, d: acc.d - val };
        if (cmd === 'up') return { ...acc, d: acc.d + val };
        return acc;
    }, { h: 0, d: 0 });

    return Math.abs(h * d);
};

/**
 * @example part2([ 'forward 5', 'down 5', 'forward 8', 'up 3', 'down 8', 'forward 2' ])
 * //=> 900
 */
const part2 = (input) => {
    const ops = input.map(o => parseOp(o));
    const { h, d } = ops.reduce((acc, [cmd, val]) => {
        if (cmd === 'forward') return { ...acc, h: acc.h + val, d: acc.d + (val * acc.a) };
        if (cmd === 'down') return { ...acc, a: acc.a + val };
        if (cmd === 'up') return { ...acc, a: acc.a - val };
        return acc;
    }, { h: 0, d: 0, a: 0 });

    return Math.abs(h * d);
};

module.exports = {
    run,
    part1,
    part2
};
